File: cpp/base/src/main/vsdk/toolkit/environment/geometry/element/Triangle.cpp
#include <cmath>
#include <cstdio>

#include "java/lang/Math.h"
#include "java/lang/String.h"
#include "vsdk/toolkit/common/VSDK.h"
#include "vsdk/toolkit/processing/Containment.h"
#include "vsdk/toolkit/environment/geometry/element/Ray.h"
#include "vsdk/toolkit/environment/geometry/element/Triangle.h"

static Vector3Dd closestPointOnTriangle(const Vector3Dd& p, const Vector3Dd& a, const Vector3Dd& b, const Vector3Dd& c)
{
    Vector3Dd ab = b.subtract(a), ac = c.subtract(a), ap = p.subtract(a);
    double d1 = ab.dotProduct(ap), d2 = ac.dotProduct(ap);
    if (d1 <= 0 && d2 <= 0) return a;
    Vector3Dd bp = p.subtract(b);
    double d3 = ab.dotProduct(bp), d4 = ac.dotProduct(bp);
    if (d3 >= 0 && d4 <= d3) return b;
    double vc = d1*d4 - d3*d2;
    if (vc <= 0 && d1 >= 0 && d3 <= 0) return a.add(ab.multiply(d1/(d1-d3)));
    Vector3Dd cp = p.subtract(c);
    double d5 = ab.dotProduct(cp), d6 = ac.dotProduct(cp);
    if (d6 >= 0 && d5 <= d6) return c;
    double vb = d5*d2 - d1*d6;
    if (vb <= 0 && d2 >= 0 && d6 <= 0) return a.add(ac.multiply(d2/(d2-d6)));
    double va = d3*d6 - d5*d4;
    if (va <= 0 && (d4-d3) >= 0 && (d5-d6) >= 0) return b.add(c.subtract(b).multiply((d4-d3)/((d4-d3)+(d5-d6))));
    double denom = 1.0/(va+vb+vc);
    double v = vb*denom, w = vc*denom;
    return a.add(ab.multiply(v)).add(ac.multiply(w));
}

Intersection* Triangle::doIntersectionWithTriangle(const Ray& ray, const Vector3Dd& v0, const Vector3Dd& v1, const Vector3Dd& v2)
{
    Vector3Dd e1 = v1.subtract(v0), e2 = v2.subtract(v0), h = ray.getDirection().crossProduct(e2);
    double a = e1.dotProduct(h);
    if (std::abs(a) < VSDK::EPSILON) return 0;
    double f = 1.0/a;
    Vector3Dd s = ray.getOrigin().subtract(v0);
    double u = f * s.dotProduct(h);
    if (u < 0 || u > 1) return 0;
    Vector3Dd q = s.crossProduct(e1);
    double v = f * ray.getDirection().dotProduct(q);
    if (v < 0 || u + v > 1) return 0;
    double t = f * e2.dotProduct(q);
    if (t <= VSDK::EPSILON) return 0;

    Vector3Dd point = ray.getOrigin().add(ray.getDirection().multiply(t));
    Vector3Dd normal = e1.crossProduct(e2).normalized();
    return new Intersection(t, point, normal);
}

int Triangle::containmentTest(const Vector3Dd& p0, const Vector3Dd& p1, const Vector3Dd& p2, const Vector3Dd& p, double distanceTolerance)
{
    Vector3Dd q = closestPointOnTriangle(p, p0, p1, p2);
    if (q.subtract(p).length() <= distanceTolerance) return static_cast<int>(Containment::LIMIT);
    return static_cast<int>(Containment::OUTSIDE);
}

void Triangle::minMax(const Vector3Dd& p0, const Vector3Dd& p1, const Vector3Dd& p2, double mm[6])
{
    mm[0] = java::Math::min(p0.x(), java::Math::min(p1.x(), p2.x()));
    mm[1] = java::Math::min(p0.y(), java::Math::min(p1.y(), p2.y()));
    mm[2] = java::Math::min(p0.z(), java::Math::min(p1.z(), p2.z()));
    mm[3] = java::Math::max(p0.x(), java::Math::max(p1.x(), p2.x()));
    mm[4] = java::Math::max(p0.y(), java::Math::max(p1.y(), p2.y()));
    mm[5] = java::Math::max(p0.z(), java::Math::max(p1.z(), p2.z()));
}

java::String Triangle::intToStr(int val) {
    char buf[32];
    snprintf(buf, sizeof(buf), "%d", val);
    return java::String(buf);
}

java::String Triangle::toString() const
{
    return java::String("f < ") + intToStr(p0) + ", " +
        intToStr(p1) + ", " +
        intToStr(p2) + " >";
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/element/Triangle.ts
import { FundamentalEntity } from "../../../common/FundamentalEntity.js";
import { Vector3Dd } from "../../../common/linealAlgebra/Vector3Dd.js";
import { VSDK } from "../../../common/VSDK.js";
import { Containment } from "../../../processing/Containment.js";
import { Ray } from "./Ray.js";
import { Intersection } from "./Intersection.js";
export class Triangle extends FundamentalEntity {
    private normal = new Vector3Dd();
    public constructor(
        private point0 = 0,
        private point1 = 0,
        private point2 = 0,
    ) {
        super();
    }
    public getPoint0() {
        return this.point0;
    }
    public getPoint1() {
        return this.point1;
    }
    public getPoint2() {
        return this.point2;
    }
    public setPoint0(value: number) {
        this.point0 = value;
    }
    public setPoint1(value: number) {
        this.point1 = value;
    }
    public setPoint2(value: number) {
        this.point2 = value;
    }
    public getNormal() {
        return this.normal;
    }
    public setNormal(x: Vector3Dd) {
        this.normal = x;
    }
    public static doIntersectionWithTriangle(ray: Ray, a: Vector3Dd, b: Vector3Dd, c: Vector3Dd): Intersection | null {
        const e1 = b.subtract(a),
            e2 = c.subtract(a),
            h = ray.getDirection().crossProduct(e2),
            d = e1.dotProduct(h);
        if (Math.abs(d) < VSDK.EPSILON) return null;
        const f = 1 / d,
            s = ray.getOrigin().subtract(a),
            u = f * s.dotProduct(h);
        if (u < 0 || u > 1) return null;
        const q = s.crossProduct(e1),
            v = f * ray.getDirection().dotProduct(q);
        if (v < 0 || u + v > 1) return null;
        const t = f * e2.dotProduct(q);
        return t <= VSDK.EPSILON
            ? null
            : new Intersection(
                  t,
                  ray.getOrigin().add(ray.getDirection().multiply(t)),
                  e1.crossProduct(e2).normalized(),
              );
    }
    public static containmentTest(a: Vector3Dd, b: Vector3Dd, c: Vector3Dd, p: Vector3Dd, t: number): number {
        const ab = b.subtract(a),
            ac = c.subtract(a),
            ap = p.subtract(a),
            d1 = ab.dotProduct(ap),
            d2 = ac.dotProduct(ap),
            denom = ab.dotProduct(ab) * ac.dotProduct(ac) - ab.dotProduct(ac) ** 2;
        if (Math.abs(denom) < VSDK.EPSILON) return Containment.OUTSIDE;
        const u = (d1 * ac.dotProduct(ac) - d2 * ab.dotProduct(ac)) / denom,
            v = (d2 * ab.dotProduct(ab) - d1 * ab.dotProduct(ac)) / denom,
            q = a.add(ab.multiply(Math.max(0, Math.min(1, u)))).add(ac.multiply(Math.max(0, Math.min(1, v))));
        return q.subtract(p).length() <= t ? Containment.LIMIT : Containment.OUTSIDE;
    }
    public static minMax(a: Vector3Dd, b: Vector3Dd, c: Vector3Dd, m: Float64Array | number[]): void {
        m[0] = Math.min(a.x(), b.x(), c.x());
        m[1] = Math.min(a.y(), b.y(), c.y());
        m[2] = Math.min(a.z(), b.z(), c.z());
        m[3] = Math.max(a.x(), b.x(), c.x());
        m[4] = Math.max(a.y(), b.y(), c.y());
        m[5] = Math.max(a.z(), b.z(), c.z());
    }
    public override toString() {
        return `f < ${this.point0}, ${this.point1}, ${this.point2} >`;
    }
}
